pkg/apis/v1beta1/nodepool.go

```diff
@@ -222,30 +239,45 @@ func (nl *NodePoolList) OrderByWeight() {
 	})
 }
 
-// MustGetAllowedDisruptions calls GetAllowedDisruptions and returns 0 if the error is not nil. This reduces the
-// amount of state that the disruption controller must reconcile, while allowing the GetAllowedDisruptions()
-// to bubble up any errors in validation.
-func (in *NodePool) MustGetAllowedDisruptions(ctx context.Context, c clock.Clock, numNodes int) int {
-	val, err := in.GetAllowedDisruptions(ctx, c, numNodes)
-	if err != nil {
-		return 0
+// GetAllowedDisruptionsByMethod returns the minimum allowed disruptions across all disruption budgets, for all disruption methods for a given nodepool
+func (in *NodePool) GetAllowedDisruptionsByReason(ctx context.Context, c clock.Clock, numNodes int) (map[DisruptionReason]int, error) {
+	allowedDisruptions := map[DisruptionReason]int{}
+	updateAllowedDisruptions := func(reason DisruptionReason, val int) {
+		if existingVal, ok := allowedDisruptions[reason]; ok {
+			allowedDisruptions[reason] = lo.Min[int]([]int{val, existingVal})
+			return
+		}
+		allowedDisruptions[reason] = val
 	}
-	return val
-}
 
-// GetAllowedDisruptions returns the minimum allowed disruptions across all disruption budgets for a given node pool.
-// This will return an error if there is a configuration error with any budget's node or schedule values.
-func (in *NodePool) GetAllowedDisruptions(ctx context.Context, c clock.Clock, numNodes int) (int, error) {
-	minVal := math.MaxInt32
-	var multiErr error
-	for i := range in.Spec.Disruption.Budgets {
-		val, err := in.Spec.Disruption.Budgets[i].GetAllowedDisruptions(c, numNodes)
+	for _, budget := range in.Spec.Disruption.Budgets {
+		val, err := budget.GetAllowedDisruptions(c, numNodes)
 		if err != nil {
-			multiErr = multierr.Append(multiErr, err)
+			return nil, err
+		}
+		if val == math.MaxInt32 {
+			// If the budget is not active, it returns maxInt32.
+			// we do not want to add non-active budgets to the allowedDisruptions map.
+			// instead we want to have these actions take the budget of the minimum active default budget
+			continue
+		}
```

Comment:
I'm wondering why we can't just add the maxInt32 in here? It seems weird to selectively skip the value, just to add it back in later. If we're comparing it to other values each time, then we know that this will be the minimum value after we iterate through all the budgets.

Reply:
you are right, i will rewrite this whole `GetAllowedDisruptionsByReason` function here as it got sloppy.